src/orquestra/sdk/_base/_traversal.py

```diff
@@ -73,26 +73,86 @@ def _make_artifact_id(source_task: model.TaskDef, future_index: int):
     )
 
 
-def _iter_futures(root_future) -> t.Iterator[_dsl.ArtifactFuture]:
-    traversal_list = [root_future]
-    traversed_nodes = set()
-    while traversal_list:
-        current_node = traversal_list.pop()
-        # using ID as all nodes are within lifetime in this function
-        # and this allows us to quickly remove duplicate iterations
-        traversed_nodes.add(id(current_node))
-        invocation = current_node.invocation
-        for arg in [*invocation.args, *[arg for _, arg in invocation.kwargs]]:
-            if isinstance(arg, _dsl.ArtifactFuture) and id(arg) not in traversed_nodes:
-                traversal_list.append(arg)
+GraphNode = t.Union[_dsl.ArtifactFuture, _dsl.Constant, _dsl.Secret]
+
+
+class GraphTraversal:
+    def __init__(self):
+        self._artifacts = {}
+        self._invocations = {}
+        self._secrets = {}
+        self._constants = {}
+
+    def traverse(self, root_futures: t.Sequence[GraphNode]):
+        """
+        Traverse the workflow graph.
+
+        We iterate over the futures returned from the workflow find the artifacts,
+        constants, and secrets.
+        """
+        node_counts: t.Dict[str, int] = defaultdict(int)
```

Comment:
**non-blocking,suggestion**: looks like there's a well-known number of entries in `node_counts`. It would make it much more readable for me if this code used just three counter variables named explicitly:
```python
artifact_counter = 0
secret_counter = 0
constant_counter = 0
```
and then:
```python
self._artifacts[_make_key(n)] = _make_artifact_node(artifact_counter, n)
```
and:
```python
artifact_counter += 1
```